src/Compilers/Core/Portable/Collections/SmallDictionary.cs

```diff
@@ -263,41 +260,26 @@ private void Insert(int hashCode, K key, V value, bool add)
             // either way nodes above unbalanced do not change their balance
             for (;;)
             {
-                // schedule hk read 
                 var hc = currentNode.HashCode;
-
-                if (currentNode.Balance != 0)
+                if (hc == hashCode)
                 {
-                    unbalancedParent = currentNodeParent;
-                    unbalanced = currentNode;
+                    this.HandleInsert(ref currentNode, key, value, add);
+                    return;
                 }
 
-                if (hc > hashCode)
-                {
-                    if (currentNode.Left == null)
-                    {
-                        currentNode.Left = currentNode = new AvlNode(hashCode, key, value);
-                        break;
-                    }
+                currentNode = ref (hc > hashCode) ?
+                                        ref currentNode.Left:
+                                        ref currentNode.Right;
 
-                    currentNodeParent = currentNode;
-                    currentNode = currentNode.Left;
-                }
-                else if (hc < hashCode)
+                if (currentNode == null)
                 {
-                    if (currentNode.Right == null)
-                    {
-                        currentNode.Right = currentNode = new AvlNode(hashCode, key, value);
-                        break;
-                    }
-
-                    currentNodeParent = currentNode;
-                    currentNode = currentNode.Right;
+                    currentNode = new AvlNode(hashCode, key, value);
```

Comment:
I like this... this is how I usually implement things in C++, it simplifies the code a lot and also removes a special case for inserting the root node.

Reply:
> also removes a special case for inserting the root node

would that be possible here by restructuring the loop and moving the null check?